plugin/restore/mysql/mysql.go

```diff
@@ -457,87 +480,183 @@ func getPITROldDatabaseName(database string, suffixTs int64) string {
+	binlogFilesOnServerSorted, err := r.GetSortedBinlogFilesMetaOnServer(ctx)
+	if err != nil {
+		return err
+	}
+	if len(binlogFilesOnServerSorted) == 0 {
+		// No binlog files on server, so there's nothing to download.
+		return nil
 	}
+	latestBinlogFileOnServer := binlogFilesOnServerSorted[len(binlogFilesOnServerSorted)-1]
+	log.Debug("Got sorted binlog file list on server", zap.Array("list", ZapBinlogFiles(binlogFilesOnServerSorted)))
 
-	todo := make(map[string]bool)
+	downloadFileList, binlogFilesLocalValidSorted, err := getBinlogDownloadListAndLocalValidFiles(r.binlogDir, binlogFilesLocal, binlogFilesOnServerSorted)
+	if err != nil {
+		return err
+	}
 
-	for _, serverBinlog := range binlogFilesOnServer {
-		// We don't download the latest binlog in SyncArchivedBinlogFiles()
-		if serverBinlog.Name == latestBinlogFileOnServer.Name {
-			continue
+	// Find if some local binlog file's first binlog event ts is already larger than targetTs.
+	// If so, we do not need to download the following binlog files on server.
+	var foundLocalFileLTTargetTs bool
+	var localFileExceedsTargetTs BinlogFile
+	for _, file := range binlogFilesLocalValidSorted {
+		eventTs, err := r.parseFirstLocalBinlogEventTimestamp(ctx, file.Name)
+		path := filepath.Join(r.binlogDir, file.Name)
+		if err != nil {
+			log.Error("Failed to parse the first binlog event timestamp for local binlog file",
+				zap.String("path", path),
+				zap.Error(err))
+			return fmt.Errorf("failed to parse the first binlog event timestamp for local binlog file[%s], error[%w]", path, err)
 		}
-
-		localBinlogSize, ok := localFileMap[serverBinlog.Name]
-		if !ok {
-			todo[serverBinlog.Name] = true
-			continue
+		if eventTs > targetTs {
+			log.Info("Found a local binlog file whose first event's timestamp is larger than targetTs",
+				zap.Any("binlogFile", file),
+				zap.Int64("targetTs", targetTs),
+				zap.Int64("eventTs", eventTs))
+			foundLocalFileLTTargetTs = true
+			localFileExceedsTargetTs = file
+			break
 		}
+	}
 
-		if localBinlogSize != serverBinlog.Size {
-			// exist on local and file size not match, delete and then download it
-			log.Warn("inconsistent binlog file size detected", zap.String("binlogFile", serverBinlog.Name), zap.Int64("server", serverBinlog.Size), zap.Int64("local", localBinlogSize))
-			if err := os.Remove(filepath.Join(r.binlogDir, serverBinlog.Name)); err != nil {
-				return fmt.Errorf("cannot remove %s, error: %w", serverBinlog.Name, err)
-			}
-			todo[serverBinlog.Name] = true
+	// Now it's time to download binlog files up to targetTs.
+	log.Debug("Download binlog file list", zap.Array("list", ZapBinlogFiles(downloadFileList)))
+	for _, file := range downloadFileList {
+		log.Debug("Downloading file", zap.Any("binlogFile", file))
+		// The about-to-download binlog file has eventTs > targetTs, so we should stop downloading the following binlog files.
+		if foundLocalFileLTTargetTs && localFileExceedsTargetTs.Seq <= file.Seq {
```

Comment:
I can get the reason for designing getBinlogDownloadListAndLocalValidFiles and having the check here to make the code more condensed. But the code is a bit hard to follow as we are trying to dealing with possible local file inconsistency and gaps as well as fetching following remote binlog files at the same time. 

I would propose a different way to make the code look more linear.

1. Get the local sorted file list
2. Get the remote sorted file list
3. Go through the existing local file list, re-download any inconsistent files and fill any gaps. If we reach targetTs during the iteration. Stop and return.
4. Go through the remote file list, download files until we reach targetTs and return.

Reply:
Filling the gap is not quite necessary, and will complicate the code (must generate filename from seq and deal with things like `binlog.999999, binlog.1000000` and binlog file basename change). So I prefer just re-download inconsistent files in step 3.